src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -759,17 +753,32 @@ private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, M
                 throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, nameof(SslStream.ReadAsync), "read"));
             }
 
+            ThrowIfExceptionalOrNotAuthenticated();
+
             Debug.Assert(_internalBuffer is null || _internalBufferCount > 0 || _decryptedBytesCount > 0, "_internalBuffer allocated when no data is buffered.");
 
             try
             {
-                while (true)
+
+                if (_decryptedBytesCount != 0)
                 {
-                    if (_decryptedBytesCount != 0)
-                    {
-                        return CopyDecryptedData(buffer);
-                    }
+                    int length = CopyDecryptedData(buffer);
+                    ReturnReadBufferIfEmpty();
+                    return length;
```

Comment:
We don't want to return here, right? We want to continue to process any additional frames we have buffered.

Reply:
we probably can. We would be here only if caller did not provide big enough buffer in previous read call. 
I can probably add check for remaining size and also do that if we can move forward without additional IO.